kymatio/scattering1d/core/timefrequency_scattering.py

```diff
@@ -64,7 +64,7 @@ def timefrequency_scattering(
     # Apply low-pass filtering over frequency (optional) and unpad
     if average:
         # zero-pad along frequency, map to Fourier domain
-        total_height = 2 ** sc_freq.J_pad
+        total_height = 2 ** sc_freq.J_pad[-1]  # TODO
```

Comment:
in light of `compute_padding_fr`, i'm pretty sure that this `# TODO` is now done

Reply:
[Nope](https://github.com/kymatio/kymatio/pull/674#issuecomment-812808596)

Reply:
ah, i see, i had missed that earlier. i'm fine with either implementation, with some preference towards padding `S1` the same as `Y2` as it will avoid duplicating the same code across backends. i don't think that this is critical for performance, is it?

an important aspect to keep in mind is that it would be good if we could disable frequential averaging on S1. (i.e., average U1 in time only and average U2 both in time and frequency)
In this way, S1 would be frequency-equivariant and S2 would be frequency-invariant

Reply:
If we don't address it, it _will_ error for some input combinations - and frankly it'd not add much code.

On averaging U1, currently this only _adds_ information, as we append unaveraged `S1`. Looks to me like the zeroth-order of `S2`, but per my latest comment unsure what'd replace it if we don't average

Reply:
fine. well, clearly we don't want unanticipated errors so the problem has to be fixed, no matter the cost in complexity.

> On averaging U1, currently this only adds information, as we append unaveraged S1. Looks to me like the zeroth-order of S2, but per my latest comment unsure what'd replace it if we don't average

I don't understand. What i had in mind was a keyword `average_S1_fr=True`, which if set to `False`, would disable frequential averaging at the first order. Something like
`average_S1_fr==True` ==> `S1 = |x * psi| * phi_t * phi_fr`
`average_S1_fr==False` ==> `S1 = |x * psi| * phi_t`

In essence, the first order of JTFS in the case `average_S1_fr=False` would fall back to the first-order of a "plain" temporal `Scattering1D`.
(feature requested by Stéphane Mallat)

Reply:
S. Mallat keeps an eye on Kymatio? My regards, he's [how I found](https://youtu.be/4eyUReyIPXg) Kymatio and implemented much of `ssqueezepy`.

I'm saying this kwarg would only save compute (by a negligible amount), since we [already append](https://github.com/lostanlen/kymatio/blob/e7b04a158ef09e2efda8b529c2ddb173c1669139/kymatio/scattering1d/core/timefrequency_scattering.py#L62) `|x * psi| * phi_t`. Or do I misunderstand?

Reply:
> S. Mallat keeps an eye on Kymatio?
yes. Glad that you got to know us via his talks!


> since we already append |x * psi| * phi_t.
oops! i hadn't seen that. it's a bit debatable to include both `S_1` and `S1_fr` though.
In their textbook definition, scattering transforms preserve energy and never duplicate it. So from that logic we should either have `S1` (averaged in time) or `S1_fr` (averaged in both time and frequency). The risk of returning both is that users might view them as different features, whereas really one can be obtained deterministically from the other.
Perhaps a solution is to have an `average_fr=True` argument which controls whether `phi_fr` is applied on _both_ `S1` and `S2`? i guess it's easier to explain to newcomers: `average` controls temporal invariance wheras `average_fr` controls frequential invariance.

At this piont, this is more an API question than an implementation question ... there's a design principle called "YAGNI" or "You Ain't Going to Need It", meaning it's important to not anticipate too many features that no one will use. But at the same time i can definitely see a use case for disabling frequential averaging (at least at the first order)

Reply:
i'm not saying that we need to implement all of this right away but it's worth to remember these aspects when designing the v1 of JTFS in Kymatio

Reply:
The averaging adds frequential invariance for `S_1` coeffs, but we already have a superior version of that in `S_2`; if it doesn't complement `S_1` any other way and add duplication, I agree it should skip.

So with `average_fr` our `psi_t * psi_f` pairs would be `|psi_t * psi_f| * phi_t`? Doable, whether a good idea I wouldn't know. But doubt we should tie it to `S1_fr`, as we might want (in fact it seems the main use case) `phi_t * phi_f` without `S1_fr`. But that means two extra kwargs...

Whatever we do, there's significant API advantage to keeping output indexed the same, even if it's mostly empty: any other code that tries to work with JTFS's output will have much extra bookkeeping to do. We'll also need all the kwargs to fetch `meta`. For visuals alone it's extra `if-else`s and indexing corrections.

Reply:
Regarding (actual) zeroth-order, I don't think we should scrap it, as we do lose significant information, especially for multi-channel data (dc bias and very low frequencies). Note this information isn't recovered anywhere else.

Reply:
right, it's true that zeroth-order has a form of information (namely, frequencies below sr/2**J) that isn't present in any other order. that include DC.
i have an audio processing bias, where the DC is irrelevant to auditory perception; but you might argue otherwise for other signal modalities like EEG.
if we can document it and it doesn't interfere too much with JTFS display routines (which assume TF inputs), then by all means let's keep the zeroth order